### ISSUE TITLE
Core/gameobject: Gossip hello is triggered twice

### ISSUE THREAD

**Issue by Killyana** [TrinityCore/TrinityCore#21678]
**Description:** The event gossip hello is triggered twice if called from a gameobject.

**Current behavior:** When the event gossip hello is triggered from a gob it will fire twice.

I'm not sure if this is related to SAI only or gossips in general.

**Steps to reproduce the problem:**

1. Execute this sql:
``` sql
UPDATE gameobject_template SET AIName="SmartGameObjectAI" WHERE entry=181147;
DELETE FROM smart_scripts WHERE entryorguid=181147 AND source_type=1;
INSERT INTO `smart_scripts` (`entryorguid`, `source_type`, `id`, `link`, `event_type`, `event_phase_mask`, `event_chance`, `event_flags`, `event_param1`, `event_param2`, `event_param3`, `event_param4`, `action_type`, `action_param1`, `action_param2`, `action_param3`, `action_param4`, `action_param5`, `action_param6`, `target_type`, `target_param1`, `target_param2`, `target_param3`, `target_x`, `target_y`, `target_z`, `target_o`, `comment`) VALUES
(181147, 1, 0, 0, 64, 0, 100, 0, 0, 0, 0, 0, 72, 0, 0, 0, 0, 0, 0, 7, 0, 0, 0, 0, 0, 0, 0, 'Gob - GOSSIP_HELLO - CLOSE_GOSSIP'),
(181147, 1, 1, 0, 64, 0, 100, 0, 0, 0, 0, 0, 12, 322, 2, 60000, 0, 0, 0, 7, 0, 0, 0, 0, 0, 0, 0, 'Gob - GOSSIP_HELLO - SUMMON_CREATURE');
```
2. .Gob add temp 181147
3. Interact with the gob once
4. notice 2 npcs are spawned instead of one

**Branch(es):** 3.3.5
**TC rev. hash/commit:** rev. 7c5b69d18bb3

**Comment by Foereaper** [comment]
This can be experienced in other places as well, as an example Ahn Kahet, the old Kingdom, where you have to press the orbs to spawn the second boss. When clicking on the object, he'll say his yell announcement twice. Another one is during the midsummer festival boss battle where you talk to an object to start the boss fight, it'll trigger the announcement twice.

**Comment by Killyana** [comment]
So this is not related to SAI only: Could be a regression of https://github.com/TrinityCore/TrinityCore/issues/19819

**Comment by Gooyeth** [comment]
@Foereaper: I can not confirm it in 6939f4eb7b023a6b2f8940ab12484523e1904860

**Comment by Killyana** [comment]
```
<Warpten> that's because gameobjects send gameobject_use and report_use
<Warpten> and/or
<Warpten> turns out that one sends both
<Warpten> so
<Warpten> GAMEOBJECT_USE is sent by CGGameObject_C::Use
<Warpten> _REPORT_USE is sent by subtype classes in client
<Warpten> scratch all i said
<Warpten> CMSG_GAME_OBJ_REPORT_USE is sent always on use (unless gameobject is "locked" - player doesnt have a spell that can open the associated lock). its not sent if the gameobject has a spell to cast on click.
<Warpten> CMSG_GAME_OBJ_USE*
<Warpten> CMSG_GAME_OBJ_REPORT_USE is only sent if using a gameobject for which player has minimum level (my names are incomplete, cant tell what its checking against) and if the subtype's CanUseNow returns true
```

**Comment by Keader** [comment]
was not fixed?

**Comment by Killyana** [comment]
There's a PR, but I not sure it's the correct way to fix it, I don't like changing current scripts.

**Comment by Keader** [comment]
I mean, ariel fixed something like that close of (https://github.com/TrinityCore/TrinityCore/commit/f913f3bb8977c127d200d5d4a608ab434b21bbcd)
no?

**Comment by Gooyeth** [comment]
@Keader: Yes, and that Ariel commit has caused this issue.

I did a pull, but I would not know the magnitude of changes to make in the DB (#21779 ). I did it based on this https://github.com/TrinityCore/TrinityCore/commit/b490d3f5ad6ccdfd103c8f756d2912f05b1741a0, assuming we have to make the changes manually.

**Comment by Langerz82** [comment]
Nedd to check if anything else breaks as a result of disabling OnReportUse in HandleGameobjectReportUse.

```diff
diff --git a/src/server/game/Handlers/SpellHandler.cpp b/src/server/game/Handlers/SpellHandler.cpp
index a26574d787..c7c206f1de 100644
--- a/src/server/game/Handlers/SpellHandler.cpp
+++ b/src/server/game/Handlers/SpellHandler.cpp
@@ -318,8 +318,8 @@ void WorldSession::HandleGameobjectReportUse(WorldPacket& recvPacket)
 
     if (GameObject* go = GetPlayer()->GetGameObjectIfCanInteractWith(guid))
     {
-        if (go->AI()->OnReportUse(_player))
-            return;
+        //if (go->AI()->OnReportUse(_player))
+        //    return;
 
         _player->UpdateAchievementCriteria(ACHIEVEMENT_CRITERIA_TYPE_USE_GAMEOBJECT, go->GetEntry());
     }
```

**Comment by jackpoz** [comment]
@Langerz82 that hack above makes no sense and just breaks an existing working functionality.

The proper way to "fix" the issue is to use event_param1 handled as "gossipHello.noReportUse"
https://github.com/TrinityCore/TrinityCore/blob/63c7bccb5014c7b356b07ab448e6b845f76ca7b2/src/server/game/AI/SmartScripts/SmartScript.cpp#L2995

When set to 1, the action is skipped when SmartGameObjectAI::OnReportUse() is called. Changing the repro sql script above adding this line makes only 1 npc spawn
```sql
UPDATE `smart_scripts` SET `event_param1` = 1 WHERE `entryorguid` = 181147 AND `id` = 1;
```
(note: setting `event_param1` = 1 to the "close gossip" action will still show the gossip as ReportUse() will open it again)

The parameter was added in https://github.com/TrinityCore/TrinityCore/commit/b0ae5fadd19fd172ec5154cde4f4fd14aa20ff88
>- SMART_EVENT_GOSSIP_HELLO - added possibility to detect for gameobject reportUse call

https://github.com/TrinityCore/TrinityCore/commit/b0ae5fadd19fd172ec5154cde4f4fd14aa20ff88#diff-55afa6409f4613551dd9ea30b26e4cdaR3301

Summing up:
- event_param1 = 0 : execute the action in both SmartGameObjectAI::GossipHello() and SmartGameObjectAI::OnReportUse()
- event_param1 = 1 : execute the action only in SmartGameObjectAI::GossipHello()

If needed it might be useful to have a 3rd option like "event_param1 = 2 : execute the action only in SmartGameObjectAI::OnReportUse()", please @Killyana let me know if that is needed.

**Comment by Killyana** [comment]
I don't really understand when  OnReportUse() is triggered.

**Comment by jackpoz** [comment]
https://github.com/TrinityCore/TrinityCore/commit/9e58ef6940c78d92f1553cf7ab75e87a6e2cf112
